File: internal/moleman/expr_test.go
package moleman

import "testing"

func TestEvalConditionBasic(t *testing.T) {
	data := map[string]any{
		"outputs": map[string]any{
			"review_json": map[string]any{
				"structured_output": map[string]any{
					"must_fix_count": 0,
				},
			},
			"previous": "ok",
		},
		"last": "ok",
	}

	cases := []struct {
		expr string
		want bool
	}{
		{"outputs.review_json.structured_output.must_fix_count == 0", true},
		{"outputs.review_json.structured_output.must_fix_count != 0", false},
		{"last == \"ok\"", true},
		{"true && false", false},
		{"true || false", true},
		{"{{ outputs.review_json.structured_output.must_fix_count == 1 }}", false},
	}

	for _, tc := range cases {
		got, err := EvalCondition(tc.expr, data)
		if err != nil {
			t.Fatalf("unexpected error for %q: %v", tc.expr, err)
		}
		if got != tc.want {
			t.Fatalf("expr %q = %v, want %v", tc.expr, got, tc.want)
		}
	}
}

func TestEvalConditionErrors(t *testing.T) {
	data := map[string]any{
		"outputs": map[string]any{
			"bad": "zero",
			"obj": map[string]any{
				"value": "nope",
			},
		},
	}

	cases := []string{
		"",
		"outputs.bad == 1",
		"outputs.obj == \"zero\"",
	}

	for _, expr := range cases {
		if _, err := EvalCondition(expr, data); err == nil {
			t.Fatalf("expected error for %q", expr)
		}
	}
}

File: test/expr.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import { evalCondition } from "../src/expr.js";

test("evalCondition handles basic expressions", () => {
  const data = {
    outputs: {
      review_json: {
        structured_output: {
          must_fix_count: 0,
        },
      },
      previous: "ok",
    },
    last: "ok",
  };

  const cases: Array<{ expr: string; want: boolean }> = [
    { expr: "outputs.review_json.structured_output.must_fix_count == 0", want: true },
    { expr: "outputs.review_json.structured_output.must_fix_count != 0", want: false },
    { expr: 'last == "ok"', want: true },
    { expr: "true && false", want: false },
    { expr: "true || false", want: true },
    { expr: "{{ outputs.review_json.structured_output.must_fix_count == 1 }}", want: false },
  ];

  for (const tc of cases) {
    assert.equal(evalCondition(tc.expr, data), tc.want, tc.expr);
  }
});

test("evalCondition returns false for missing and non-matching expressions", () => {
  const data = {
    outputs: {
      bad: "zero",
      obj: {
        value: "nope",
      },
    },
  };

  assert.equal(evalCondition("outputs.missing == 1", data), false);
  assert.equal(evalCondition("outputs.bad == 1", data), false);
  assert.equal(evalCondition('outputs.obj == "zero"', data), false);

  assert.throws(() => evalCondition("", data));
});
